### DISCUSSION TITLE
latest/trackers/

### DISCUSSION THREAD

**Discussion by giscus** [roboflow/supervision#950] [Docs]
# latest/trackers/

A set of easy-to-use utilities that will come in handy in any computer vision project.

https://supervision.roboflow.com/latest/trackers/

**Comment by FrancisDrakeII** [comment]
in "update_with_detections(detections)" example, "tracker = sv.ByteTrack()" "detections = byte_tracker.update_with_detections(detections)" should be "detections = tracker.update_with_detections(detections)"

**Reply by SkalskiP** [reply]
Thanks for letting us know! Fixed!

**Comment by s-madhavan001** [comment]
how to track in each person wait time in using supervision ?
i can detect in every single people with track id. what will do next. for track **each id's idel-time(standing/waiting time)**
in using supervision or ultralytics(yolov8)

**Reply by SkalskiP** [reply]
Hi @s-madhavan001 👋🏻 All you need is here: https://github.com/roboflow/supervision/tree/develop/examples/time_in_zone

**Reply by SkalskiP** [reply]
Not really sure what you mean. It is already tracking and showing IDs.

**Reply by s-madhavan001** [reply]
have any Tutorial videos ?

**Reply by SkalskiP** [reply]
Working on it ;) It should be out within the next 2 days.

**Reply by SkalskiP** [reply]
Hi @s-madhavan001, the video was released yesterday: https://youtu.be/hAWpsIuem10. I hope you'll find it helpful. I'm sorry it took a bit longer. It's a complicated topic. :/

**Comment by Shahid-TheMatrixMaker** [comment]
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
chex 0.1.86 requires numpy>=1.24.1, but you have numpy 1.23.0 which is incompatible.
pandas-stubs 2.0.3.230814 requires numpy>=1.25.0; python_version >= "3.9", but you have numpy 1.23.0 which is incompatible.
tensorflow 2.15.0 requires numpy<2.0.0,>=1.23.5, but you have numpy 1.23.0 which is incompatible.

i found this error while installing super-gradients

**Comment by aberg2323** [comment]
I am currently testing out the bytetrack code with my yolov8 object detection model. I am annotating my video with tracers. The problem I am encountering is that sometimes the tracers are working perfect, but then it seems the model "loses" the object for a little bit, and the tracer restarts. I have already tried changing some of the parameters of bytetrack and am seeing marginally improved results, but nothing that completely fixes the problem. Is this something that is solvable by changing the parameters of bytetrack or something that would be fixed by me training my object detection model again with more data?

**Reply by rolson24** [reply]
Hi @aberg2323!

Usually when ByteTrack is not performing as expected it is due to the object detection model have less than ideal performance. The tracker relies on objects being detected in nearly every frame, and also relies heavily on the confidence score from the detection model to determine if the detection is a valid object. I would try training your object detection model more and once it seems to have good performance on the task you want it to perform, then you can adjust the parameters of ByteTrack.

**Reply by aberg2323** [reply]
That is what I figured! Thank you for the help

**Comment by ashish-reddy26** [comment]
hey @skalskiP

I am doing an analysis with object tracking, so far bot-sort performed well than Bytetrack, but I want to go with Bytetrack, as I aimed to work with videos at 30FPS.

Firstly, performed the detection and tracking using supervision. Got the results, but as I want to do lot of analysis in different parts of the frame, to do this I need to pass the detections separately by leaving out some of the labels, to achieve this, I tried the following way.

I performed detection and tracking separately (meaning 

 i. Initially, I performed detection on the video and stored the results in a variable 'custom_results'
 ii. Then the detections of each frame (custom_results[i]) were given to the tracker manually, as expected the results were same by the both approaches.)

Is there a way that I can pass the labels of required  area of interest to the tracker? (As I referred the documentation of supervision, I saw that the detections were taken as tensors.)

Can you give any leads on approaching this.

Thank you,
Ashish

**Reply by ashish-reddy26** [reply]
Thanks,
Figured it out, thanks for the 'supervision'

Ashish

**Comment by AdelShousha** [comment]
I'm tracking cars using an rtsp link, after a while a get errors such as:

```
[rtsp @ 0x5cf0efc9cf80] RTP: PT=60: bad cseq 1f2e expected=df9f
[hevc @ 0x5cf0ef87c200] The cu_qp_delta -31 is outside the valid range [-26, 25].
                            # cu_qp_delta 125
Frame resolution: 2560x1920

0: 480x640 (no detections), 4.1ms
Speed: 1.6ms preprocess, 4.1ms inference, 0.5ms postprocess per image at shape (1, 3, 480, 640)
[hevc @ 0x5cf0efb23140] Could not find ref with POC 30

[rtsp @ 0x5d018ba71180] Illegal temporal ID in RTP/HEVC packet

[hevc @ 0x5d018b974ec0] CABAC_MAX_BIN : 7
```

I'm using cv2.VideoCapture(RTSP) to capture the rtsp stream and I have these lines to check the continuity of the stream

```python
while True:
    ret, frame = cap.read()
    if not ret:
        print("No frame read")
        break
```

so after a while from using the rtsp stream it breaks  

I know this error might not be related to tracking exclusively but I'm not able to understand why is this problem happening is it something with the rtsp stream itself or my processing or the use of cv2.VideoCapture() or what exactly, can someone help?

**Reply by SkalskiP** [reply]
Hi @AdelShousha 👋🏻 Yup, the error is unrelated to tracking. Is that stream stable? Are you able to open it with other tools like VLC?

**Reply by AdelShousha** [reply]
yes the stream is stable and I'm able to watch ot using VLC without any problems

**Reply by SkalskiP** [reply]
Have you tried using `InferencePipeline`? https://inference.roboflow.com/using_inference/inference_pipeline/

**Reply by AdelShousha** [reply]
Thank you for suggesting inference!, I have tried it and the previous error is resovled but I have a new problem. My code works normally for a while then the terminal yields "Killed" stopping the process, when running "sudo dmesg | grep -i 'killed'" I get :
 [695416.057171] Out of memory: Killed process 1711614 (pt_main_thread) total-vm:23405672kB, anon-rss:6538112kB, file-rss:68640kB, shmem-rss:12160kB, UID:1000 pgtables:19644kB oom_score_adj:0

I'm using a custom inference logic like the one provided here :https://inference.roboflow.com/using_inference/inference_pipeline/#how-to-provide-a-custom-inference-logic-to-inferencepipeline
and I'm using a custom sink that uses two other yolo models on the video frame coming from my original yolo making it a total of three models running at the same time.

I don't know if the way I'm using three models is causing the problem or if something else causing this problem. Can someone help?

**Reply by SkalskiP** [reply]
@AdelShousha, let me ask someone from the Inference team.

Btw, could you be a bit more specific here? `My code works normally for a while` How long does it take for the pipeline to crash? How much RAM do you have on that machine?

**Reply by grzegorz-roboflow** [reply]
Hi @AdelShousha , this is Grzegorz here, I'm part of `inference` team.

It seems OOM kicks in and kills your script. Since `inference-pipeline` is buffering frames it is quite possible that your RAM is maxed out, can you try `export VIDEO_SOURCE_BUFFER_SIZE=2` when running the script and see if the script completes without abnormal termination?

**Comment by rsnk96** [comment]
Hi

Amazing project!

Just a note: The parameter `lost_track_buffer` of the tracker in mislabelled here in the docs as `track_buffer`.

**Comment by Muhammad-237** [comment]
are you adding other trackers like deepocsort etc in near future ?

**Comment by fenaux** [comment]
How can I know the indexes of the boxes corresponding to trackers_id ?

After detection, I get 34 boxes  class 0
After tracker update only 24 boxes are remaining

I use this trick

n_in_i_frame = len(boxes_in_i_frame)

detections = sv.Detections(xyxy = boxes_in_i_frame,
                                      class_id=np.zeros(n_in_i_frame),
                                      confidence=conf_in_i_frame,
                                      data={'0':np.arange(n_in_i_frame)})

detections = tracker.update_with_detections(detections)

and then
detections.data['0']
 array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 23, 25])

Am I right, these are the original indexes of boxes as given by np.arange(n_in_i_frame)} from 0 to 33 ?
Is there a better way to know which box as been assigned to a track_id and which one were not assigned

Suggestion : 
an other way to do that would be to keep all the boxes after tracker.update_with_detections(detections)
assigning each box to a track_id and a special track_id -1 for those not assigned to a track. that would be consistent to convention in scikilt-learn dbscan for instance.

Anyway, thanks a lot for providing all these usefull tools.

**Reply by NANOBYTECOMPUTERS** [reply]
I dont know who may be trying to figure this out but here was the way I accomplished what I did (just the relevant parts) Good luck I beleive in you ;D

    for result in results:  # Stream mode yields one result per frame
        detections = sv.Detections.from_ultralytics(result)
        return tracker.update(detections) or sv.Detections.empty()
    return sv.Detections.empty()


            result = perform_detection(model, context.tracker) or sv.Detections.empty()
            context.frame_parser.parse(result)

        if isinstance(frame, sv.Detections):
            boxes_array, classes_array, ids_array, num_detections = self._convert_sv_to_numpy(frame)
            confidence_array = frame.confidence

        self.label_annotator = sv.LabelAnnotator(
#my detection lables to display
        )

  I than converted my detections to numpy arrays

                if self.current_locked_target_id is not None:
                    tracker_ids = frame.tracker_id
                    locked_idx = self._get_locked_index(tracker_ids)
                    #for tracking and targeting 
                    if locked_idx != -1:
                        box = frame.xyxy[locked_idx]
                        center_x, center_y, cls, id = self._extract_target_info(frame, locked_idx, box)
                        current_pos = np.array([center_x, center_y])
                        prev_pos = np.array([self.previous_target.x, self.previous_target.y]) if self.previous_target else current_pos
                        
                        if np.linalg.norm(current_pos - prev_pos) <= self.switch_threshold:
                            if cls == 0:  # If the target is a player
                                # Check for associated head
                                head_targets = [t for t in zip(frame.xyxy, frame.class_id, frame.tracker_id) if t[1] == HEADSHOT_CLASS and t[2] == id]
                                if head_targets:
                                    head_target = Target(*head_targets[0][0], HEADSHOT_CLASS, head_targets[0][2])
                                    self.context.mouse.process_data((head_target.center_x, head_target.center_y, head_target.w, head_target.h, HEADSHOT_CLASS, head_target.id))
                                    self.context.visuals.draw_aim_point(head_target.center_x, head_target.center_y)
                                    self.previous_target = head_target
                                    self.current_locked_target_id = head_target.id
                                else:
                                    # No head detected, keep the player body
                                    self.context.mouse.process_data((center_x, center_y, box[2] - box[0], box[3] - box[1], cls, id))
                                    self.context.visuals.draw_aim_point(center_x, center_y)
                                    self.previous_target = Target(*box, cls, id)
                            else:
                                # Use the target data directly if not a player
                                self.context.mouse.process_data((center_x, center_y, box[2] - box[0], box[3] - box[1], cls, id))
                                self.context.visuals.draw_aim_point(center_x, center_y)
                                self.previous_target = Target(*box, cls, id)
                        else:
                            # Target has moved too far, potentially switch to a new one
                            new_target_info = Target(*box, cls, id)
                            self.context.mouse.process_data((center_x, center_y, box[2] - box[0], box[3] - box[1], cls, id))
                            self.current_locked_target_id = id
                            self.previous_target = new_target_info

**Comment by kennedy-kitoko** [comment]
can we track in the differente frame , multi frame camera  but same object ?

**Comment by kennedy-kitoko** [comment]
Hello supervision Team,

I'm a user working on object tracking in real-time environments, specifically using YOLO and DeepSORT/ByteTrack for multiple object tracking (MOT). The supervision library has been immensely helpful for single-camera (mono-camera) tracking, particularly with the seamless integration of bounding box conversion, tracking, and annotation.

The Proposed Feature: Multi-Camera Tracking (MTMC)
I would like to suggest and request the development of a feature that supports Multi-Camera Tracking (MTMC), allowing users to associate a single track ID with the same object across multiple video streams.

Motivation
Currently, the robust tracking capabilities are confined to single streams. For users deploying surveillance, traffic monitoring, or large-area people-counting systems, the biggest challenge is ID consistency when an object moves from the field of view of one camera to another.

A dedicated MTMC component would significantly increase the utility of supervision for real-world deployments.

Suggested Implementation Approach (High-Level)
The MTMC functionality would require a Global Association layer on top of the existing supervision.Detections and tracker logic. This could potentially be managed by a new class, e.g., sv.MultiCameraTracker.

Re-ID Feature Extraction: Integrate with a lightweight Person Re-Identification (Re-ID) model (like OSNet or similar) to extract appearance features from the detected bounding boxes before they are passed to the DeepSORT/ByteTrack instance.

Per-Camera Tracking: Run the standard sv.Trackers (DeepSORT/ByteTrack) on each camera stream independently to get local track IDs.

Global Association: Implement a mechanism (e.g., a simple greedy assignment or graph-based matching) that compares the Re-ID features of the lost tracks in Camera A with the newly appearing tracks in Camera B to assign a consistent Global ID.

This feature would solidify supervision as the go-to library for production-ready MOT and MTMC solutions.

Thank you for considering this request and for all your great work on the library!

**Comment by kvnptl** [comment]
Thank you for the tracking object feature on Supervision. Are there plans to add the ReID feature?

**Reply by SkalskiP** [reply]
Thank you for the question. Yes, we plan to add ReID. We were looking for someone strong enough technically to help build features like this, and we found that person. You should expect steady growth in our trackers over the next months.

**Reply by kvnptl** [reply]
That's awesome news!! Looking forward it!!🙏
